File: backend/schemas/books.py
from datetime import date
from typing import List, Optional
from pydantic import BaseModel, Field

from backend.schemas.common import RatingStatus


class Author(BaseModel):
    id: str 
    name: str


class Rating(BaseModel):
    average: float = Field(ge=0, le=5)
    count: int = Field(ge=0)
    asOf: Optional[date] = None


class Edition(BaseModel):
    id: str
    publishYear: Optional[int] = None
    isbn13: Optional[str] = None


class Book(BaseModel):
    id: str
    title: str
    authors: List[Author]
    firstPublishYear: Optional[int] = None
    coverUrl: Optional[str] = None
    rating: Optional[Rating] = None
    ratingStatus: RatingStatus


class SubjectRef(BaseModel):
    slug: str
    name: str


class BookDetail(Book):
    editionCount: Optional[int] = 0
    language: Optional[str] = None
    description: Optional[str] = None
    coverUrlLarge: Optional[str] = None
    subjects: list[SubjectRef]
    editions: list[Edition] = Field(default_factory=list, max_length=5)

File: frontend/src/api/books.js
const API_KEY = import.meta.env.VITE_APP_KEY;

async function apiFetch(url) {
    const res = await fetch(url, {
        headers: { "X-App-Key": API_KEY },
    });
    if (!res.ok) {
        throw new Error("Failed to fetch");
    }
    return res.json();
}

export async function fetchBooks(query) {
    const url = query ? `/api/v1/books/search?q=${encodeURIComponent(query)}` : `/api/v1/books/popular?limit=24`;
    const data = await apiFetch(url);
    const items = data.items || [];
    return items.map(book => ({
        id: book.id,
        title: book.title,
        author: book.authors && book.authors.length > 0 ? book.authors.map(a => a.name).join(', ') : 'Unknown Author',
        coverUrl: book.coverUrl,
        rating: book.rating && typeof book.rating === 'object' ? book.rating.average : (book.rating || "No rating"),
        description: book.description || "" 
    }));
}

export async function fetchCategories() {
    const data = await apiFetch('/api/v1/subjects/');
    return data.items || [];
}

export async function fetchBooksByCategory(slug) {
    const data = await apiFetch(`/api/v1/subjects/${slug}/books?limit=24`);
    const items = data.items || [];
    return items.map(book => ({
        id: book.id,
        title: book.title,
        author: book.authors && book.authors.length > 0 ? book.authors.map(a => a.name).join(', ') : 'Unknown Author',
        coverUrl: book.coverUrl,
        rating: book.rating && typeof book.rating === 'object' ? book.rating.average : (book.rating || "No rating"),
        description: book.description || ""
    }));
}

export async function fetchBookDetail(bookId) {
    const book = await apiFetch(`/api/v1/books/${bookId}`);
    return {
        id: book.id,
        title: book.title,
        author: book.authors && book.authors.length > 0 ? book.authors.map(a => a.name).join(', ') : 'Unknown Author', 
        coverUrl: book.coverUrlLarge || book.coverUrl,
        rating: book.rating && typeof book.rating === 'object' ? book.rating.average : (book.rating || "No rating"),
        description: book.description || "No description available.",
        firstPublishYear: book.firstPublishYear,
        editionCount: book.editionCount, 
        subjects: book.subjects || [],
        language: book.language
    };
}
